File: flutter/lib/FileBenchmark.dart
import 'dart:io';

import 'package:path_provider/path_provider.dart';

class FileBenchmark {
  static const int _times = 50;
  int _timer = 0;

  static const String FILE_NAME = "test_file";
  static const String FILE_NAME2 = "test_file2";
  static const String FILE_GENERATED = "test_file_generated";
  String _path;

  FileBenchmark() {
    getApplicationDocumentsDirectory().then((p) => _path = p.path);
  }

  Future<double> saveFile() async {
    int result = 0;

    List<int> file = _readFile(FILE_GENERATED);
    for (int i = 0; i < _times; i++) {
      _timer = DateTime.now().microsecondsSinceEpoch;
      _saveFile(FILE_NAME, file);
      result += DateTime.now().microsecondsSinceEpoch - _timer;
      _deleteFile(FILE_NAME);
    }

    return result / _times;
  }

  Future<double> readFile() async {
    int result = 0;
    int dummy = 0;

    for (int i = 0; i < _times; i++) {
      _timer = DateTime.now().microsecondsSinceEpoch;
      dummy += _readFile(FILE_GENERATED).length;
      result += DateTime.now().microsecondsSinceEpoch - _timer;
    }

    print(dummy);
    return result / _times;
  }

  Future<double> deleteFile() async {
    int result = 0;

    List<int> file = _readFile(FILE_GENERATED);
    for (int i = 0; i < _times; i++) {
      _saveFile(FILE_NAME, file);
      _timer = DateTime.now().microsecondsSinceEpoch;
      _deleteFile(FILE_NAME);
      result += DateTime.now().microsecondsSinceEpoch - _timer;
    }

    return result / _times;
  }

  void _saveFile(String name, List<int> data) {
    return File('$_path/$name').writeAsBytesSync(data);
  }

  List<int> _readFile(String name) {
    return File('$_path/$name').readAsBytesSync();
  }

  void _deleteFile(String name) {
    return File('$_path/$name').deleteSync();
  }
}

File: kotlin/app/src/commonMain/kotlin/kotlin_test/FileBenchmark.kt
package kotlin_test

class FileBenchmark(private val fileHelper: FileHelper) {

    private val times = 50
    private var timer = 0L

    private val fileSize = 100
    private val fileName = "test_file"
    private val fileName2 = "test_file2"

    fun saveFile(): Double {
        var result = 0L

        fileHelper.generateFile(fileName2, fileSize)
        val file = fileHelper.readFile(fileName2)
        repeat(times) {
            timer = Util.getNanoTime()
            fileHelper.saveFile(fileName, file)
            result += Util.getNanoTime() - timer
            fileHelper.deleteFile(fileName)
        }

        fileHelper.deleteFile(fileName2)
        return result / times.toDouble()
    }

    fun readFile(): Double {
        var result = 0L
        var dummy = 0

        fileHelper.generateFile(fileName, fileSize)
        repeat(times) {
            timer = Util.getNanoTime()
            dummy += fileHelper.readFile(fileName).size
            result += Util.getNanoTime() - timer
        }

        println(dummy)
        fileHelper.deleteFile(fileName)
        return result / times.toDouble()
    }

    fun deleteFile(): Double {
        var result = 0L
        var dummy = false

        fileHelper.generateFile(fileName2, fileSize)
        val file = fileHelper.readFile(fileName2)
        repeat(times) {
            fileHelper.saveFile(fileName, file)
            timer = Util.getNanoTime()
            dummy = fileHelper.deleteFile(fileName)
            result += Util.getNanoTime() - timer
        }

        println(dummy)
        fileHelper.deleteFile(fileName2)
        return result / times.toDouble()
    }

}

expect class FileHelper {

    fun saveFile(name: String, data: ByteArray)

    fun readFile(name: String): ByteArray

    fun deleteFile(name: String): Boolean

    fun generateFile(name: String, size: Int)

}